### PR TITLE
Revised prepare_data, _reduce and _non_zero_means to process water quality data from monitoring location

### PR DESCRIPTION
**Pull request by gcfang**
Test functions were added to test_analysis.py. All three tests passed. All three functions are ready for review

### CODE REVIEW

File: propagator/analysis.py [hunk 1]
```diff
@@ -239,8 +239,8 @@ def split_streams(stream_layer, subcatchment_layer):
 
 
 @utils.update_status()
-def prepare_data(mon_locations, monloc_id_col, subcatchments,
-                 subcatch_id_col, final_fields, outputfile,
+def prepare_data(mon_locations, subcatchments,subcatch_id_col,
```
Comment (phobson) [inline propagator/analysis.py]: Space after comma

File: propagator/analysis.py [hunk 2]
```diff
@@ -362,21 +356,77 @@ def _reduce(_ml, _out_ml):
...
+    wq_fields : list
+        Lisf of water quality parameters to be analyzed
+    subcatch_id_col : str
+        Field name of the subcatchment ID column in ``subcatchments``.
 
     Returns
     -------
     _out_ml : str
-
+
```
Comment (phobson) [inline propagator/analysis.py]: remove extra/trialing white space on this line

File: propagator/analysis.py [hunk 3]
```diff
@@ -362,21 +356,77 @@ def _reduce(_ml, _out_ml):
...
+    _sqlexp = '"{}" = {}'.format(sort_id, numpy.min(_arr[sort_id]))
     arcpy.analysis.Select(_ml, _out_ml, _sqlexp)
-
-    #process the WQ data - [e.g. out_ml.wq = average(temp_ml.wq)]
+    #_sqlexp = "%s" + '= ' + "%i" % sort_id, numpy.min(_arr[sort_id])
+    
+    
+    # Loop through each water quality parameter.
+    for wq_par in wq_fields:
+        wq_value  = utils.groupby_and_aggregate(
```
Comment (phobson) [inline propagator/analysis.py]: remove extra spaces

File: propagator/analysis.py [hunk 4]
```diff
@@ -362,21 +356,77 @@ def _reduce(_ml, _out_ml):
...
+            input_path= _ml,
+            groupfield=subcatch_id_col,
+            valuefield=wq_par,
+            # If at futre stage it is decided to compute the water quality 
+            # value with a different methods (i.e. median), either revise 
+            # the existing _non_zero_means function, or (preferrably) write
+            # a new function (i.e. non_zero_median) and call the new function
+            # at the line below.
+            aggfxn= _non_zero_means
```
Comment (phobson) [inline propagator/analysis.py]: remove spaces around `=` between the `(...)` of the function call. move comments outside of the `(...)` as well.

File: propagator/analysis.py [hunk 5]
```diff
@@ -362,21 +356,77 @@ def _reduce(_ml, _out_ml):
...
+    """
+    
+    
+    try:
+        # For some reason when this function is called from _reduce, _arr
+        # is not treated as a regular list, and only the following line
+        # will work to correctly extract numeric values from _arr
+        _numlst = filter(lambda n: n>0, [r[1] for r in _arr])
+    except:
+        # This line works when _arr is a regular list (for which case the
```
Comment (phobson) [inline propagator/analysis.py]: Don't just catch *any* error. Be  specific with something like `except IndexError` (or what ever error you're actually getting, Also, write a test that demonstrates this behavior.

Comment (phobson) [inline propagator/analysis.py]: Upon thinking about this, the statement in the `try` block is intended for arcpy Cursors. So really we should just be checking the types, there. e.g., 

```python
if isinstance(_arr, arcpy.Cursor)`:
    # yada yada
elif isinstance(_arr, numpy.ndarray):
    # yada
else:
    raise ValueError("only numpy arrays and arcpy Cursors are supported")
```

File: propagator/analysis.py (hunk 1 above)
Comment (gcfang) [inline propagator/analysis.py]: revised

File: propagator/analysis.py (hunk 2 above)
Comment (gcfang) [inline propagator/analysis.py]: I think the space is deleted now. Please verify.

File: propagator/analysis.py (hunk 3 above)
Comment (gcfang) [inline propagator/analysis.py]: removed

File: propagator/analysis.py (hunk 4 above)
Comment (gcfang) [inline propagator/analysis.py]: all spaces removed

File: propagator/analysis.py (hunk 5 above)
Comment (gcfang) [inline propagator/analysis.py]: Adopted the suggested structure to verify input data type

**Pull request merged**